python/hail/methods/misc.py

```diff
@@ -358,4 +369,185 @@ def window_by_locus(mt: MatrixTable, bp_window_size: int) -> MatrixTable:
+            j += 1
+        starts[i] = j
+
+        max_val = a[i] + radius
+        while k < size and a[k] <= max_val:
+            k += 1
+        stops[i] = k
+
+    return starts, stops
+
+
+@typecheck(locus_table=Table,
+           radius=oneof(int, float),
+           value_expr=nullable(expr_float64))
+def locus_windows(locus_table, radius, value_expr=None):
+    """Returns start and stop indices for window around each locus.
+
+    Examples
+    --------
+    Windows with 2bp radius for one contig with positions 1, 2, 3, 4, 5:
+
+    >>> mt = hl.balding_nichols_model(n_populations=1, n_samples=5, n_variants=5)
+    >>> starts, stops = hl.locus_windows(mt.rows(), 2)
+    >>> starts, stops
+    (array([0, 0, 0, 1, 2]), array([3, 4, 5, 5, 5]))
+
+    The following examples involve three contigs.
+
+    >>> loci = [{'locus': hl.Locus('1', 1), 'cm': 1.0},
+    ...         {'locus': hl.Locus('1', 2), 'cm': 3.0},
+    ...         {'locus': hl.Locus('1', 4), 'cm': 4.0},
+    ...         {'locus': hl.Locus('2', 1), 'cm': 2.0},
+    ...         {'locus': hl.Locus('2', 1), 'cm': 2.0},
+    ...         {'locus': hl.Locus('3', 3), 'cm': 5.0}]
+    >>> ht = hl.Table.parallelize(
+    ...         loci,
+    ...         hl.tstruct(locus=hl.tlocus('GRCh37'), cm=hl.tfloat64),
+    ...         key=['locus'])
+
+    Windows with 1bp radius:
+
+    >>> hl.locus_windows(ht, 1)
+    (array([0, 0, 2, 3, 3, 5]), array([2, 2, 3, 5, 5, 6]))
+
+    Windows with 1cm radius:
+
+    >>> hl.locus_windows(ht, 1.0, value_expr=ht.cm)
+    (array([0, 1, 1, 3, 3, 5]), array([1, 3, 3, 5, 5, 6]))
+
+    Notes
+    -----
+    The first key field of `locus_table` must be of type :py:data:`tlocus`, as
+    is true for variant-keyed datasets. The rows are then ordered by locus, i.e.
+    by contig and then by position within contig.
+
+    This function returns two 1-dimensional ndarrays of integers,
+    ``starts`` and ``stops``, each of size equal to the number of rows.
+
+    By default, for all indices ``i``, ``[starts[i], stops[i])`` is the maximal
+    range of row indices ``j`` such that ``contig[i] == contig[j]`` and
+    ``position[i] - radius <= position[j] <= position[i] + radius``.
+
+    Set `value_expr` to use a value other than position to define the windows.
+    This row-indexed numeric expression on `locus_table` must be non-missing,
+    as well as non-decreasing with respect to locus for each contig; otherwise
+    the function will fail.
+
+    The last example above uses centimorgan coordinates, so
+    ``[starts[i], stops[i])`` is the maximal range of row indices ``j`` such
+    that ``contig[i] == contig[j]`` and
+    ``cm[i] - radius <= cm[j] <= cm[i] + radius``.
+
+    Index ranges are start-inclusive and stop-exclusive. This function is
+    especially useful in conjunction with
+    :meth:`.BlockMatrix.sparsify_row_intervals`.
+
+    Parameters
+    ----------
+    locus_table : :class:`.Table`
+        Table with first key of type :py:data:`tlocus`.
```

Comment:
I don't understand why the first key has to be type locus. Why not just take an optional locus parameter that defaults to "locus". Similar to what i did for export_plink.